File: apps/fsharp/design-pattern/src/DesignPattern/Strategy.fs
namespace DesignPattern.Strategy

/// Strategy パターン
/// F# では関数が第一級値なので、インターフェースは不要。
/// 戦略を関数として渡すだけで実現できる。
module Strategy =

    /// フォーマット戦略の型エイリアス
    type FormatStrategy = string -> string

    /// レポートのコンテキスト
    type Report = { Title: string; Items: string list }

    /// レポートを指定された戦略でフォーマットする
    let formatReport (strategy: FormatStrategy) (report: Report) : string =
        let header = strategy report.Title
        let body = report.Items |> List.map strategy |> String.concat "\n"
        sprintf "%s\n%s" header body

    /// HTML 戦略
    let htmlStrategy: FormatStrategy = fun text -> sprintf "<p>%s</p>" text

    /// プレーンテキスト戦略
    let plainTextStrategy: FormatStrategy = fun text -> sprintf "** %s **" text

    /// Markdown 戦略
    let markdownStrategy: FormatStrategy = fun text -> sprintf "- %s" text

File: apps/haskell/design-pattern/src/Strategy.hs
-- | Strategy パターン
-- 高階関数でアルゴリズムを差し替える。
-- Haskell では関数が第一級値なので、Strategy は自然な設計。
module Strategy
  ( Formatter
  , Report(..)
  , formatReport
  , htmlFormatter
  , plainTextFormatter
  , markdownFormatter
  ) where

-- | フォーマット戦略: タイトルと行リストを受け取り文字列を返す関数
type Formatter = String -> [String] -> String

-- | レポートデータ
data Report = Report
  { reportTitle :: String
  , reportBody  :: [String]
  } deriving (Show, Eq)

-- | レポートをフォーマットする（戦略を注入）
formatReport :: Formatter -> Report -> String
formatReport fmt r = fmt (reportTitle r) (reportBody r)

-- | HTML 戦略
htmlFormatter :: Formatter
htmlFormatter title body =
  "<html><head><title>" ++ title ++ "</title></head><body>\n"
  ++ concatMap (\l -> "  <p>" ++ l ++ "</p>\n") body
  ++ "</body></html>\n"

-- | プレーンテキスト戦略
plainTextFormatter :: Formatter
plainTextFormatter title body =
  "***** " ++ title ++ " *****\n"
  ++ concatMap (\l -> l ++ "\n") body

-- | Markdown 戦略
markdownFormatter :: Formatter
markdownFormatter title body =
  "# " ++ title ++ "\n\n"
  ++ concatMap (\l -> "- " ++ l ++ "\n") body
